ta/aes_crypto_ta.c

```diff
@@ -271,6 +282,113 @@ static TEE_Result copy_secure_memory(uint32_t param_types, TEE_Param params[4])
   return TEE_SUCCESS;
 }
 
+#define CTR_AES_BLOCK_SIZE  16
+#define CTR_AES_IV_SIZE CTR_AES_BLOCK_SIZE
+#define CTR_AES_KEY_SIZE CTR_AES_BLOCK_SIZE
+
+struct sub_sample_t {
+    uint32_t clear_bytes;
+    uint32_t encrp_bytes;
+};
+
+static TEE_Result aes_Ctr128_Encrypt_secure(uint32_t param_types,
+                                     TEE_Param params[TEE_NUM_PARAMS])
+{
+  TEE_Result res = TEE_SUCCESS;
+  void *key, *iv, *inbuf, *outbuf, *iter_in, *iter_out;
+  uint32_t insz, outsz, offset = 0;
+  struct sub_sample_t *sub_samples;
+  uint8_t *samples_end;
+  uint32_t  exp_param_types = AES_CTR128_ENCRYPT_SECURE_TEE_PARAM_TYPES;
+
+  if (param_types != exp_param_types) {
+    EMSG("%s: incorrect parameters", __func__);
+    return TEE_ERROR_BAD_PARAMETERS;
+  }
+
+  if (params[0].memref.buffer == NULL || params[0].memref.size == 0)
+    return TEE_ERROR_BAD_PARAMETERS;
+
+  if (params[1].memref.buffer == NULL || params[1].memref.size == 0)
+    return TEE_ERROR_BAD_PARAMETERS;
+
+  if (params[2].memref.buffer == NULL ||
+      params[2].memref.size < sizeof(struct sub_sample_t))
+    return TEE_ERROR_BAD_PARAMETERS;
+
+  if (params[3].memref.buffer == NULL ||
+      params[3].memref.size < CTR_AES_KEY_SIZE + CTR_AES_IV_SIZE)
+    return TEE_ERROR_BAD_PARAMETERS;
+
+  inbuf = params[0].memref.buffer;
+  insz = params[0].memref.size;
+  outbuf = params[1].memref.buffer;
+  outsz = params[1].memref.size;
+  sub_samples = (struct sub_sample_t *)params[2].memref.buffer;
+  samples_end = ((uint8_t *)params[2].memref.buffer + params[2].memref.size);
+
+  /*Encrypt_secure function only be called in 'CFG_SECURE_DATA_PATH=y' case */
+  res = TEE_CheckMemoryAccessRights(TEE_MEMORY_ACCESS_ANY_OWNER |
+                                    TEE_MEMORY_ACCESS_WRITE |
+                                    TEE_MEMORY_ACCESS_SECURE,
+                                    outbuf, outsz);
+  if (res != TEE_SUCCESS) {
+    EMSG("%s: WARNING: output buffer is not in secure memory", __func__);
+    return TEE_ERROR_SECURITY;
+  }
+
+  key = params[3].memref.buffer;
+  iv = (uint8_t*)key + CTR_AES_KEY_SIZE;
+  iter_in = inbuf;
+  iter_out = outbuf;
+
+  while (sub_samples->clear_bytes != 0xFFFFFFFF &&
+         (uint8_t *)sub_samples < samples_end) {
```

Comment:
To ensure you are not accessing outside of the provided 'sub_samples' buffer:

1) Check if you reached the end of the provided buffer **before** accessing 'sub_samples->clear_bytes'.
2) Account for the size of the 'sub_sample_t' structure.

Something like:

```
while (((uint8_t *)sub_samples + sizeof(struct sub_sample_t)) <= samples_end &&
       sub_samples->clear_bytes != 0xFFFFFFFF) {
    //...
}
```

Alternatively, you could consider to compute the number of sub samples based on the size of the provided buffer and loop through the provided sub samples using a 'for' loop.

Reply:
Swap these two conditions should be OK for sub_samples checking.
sub_samples can not exceed the samples_end with this checking.

And for actual clear_bytes and encrypted_bytes, an overlap checking is already provided so it is unnecessary that we sum all bytes and checking range.

For sub_samples